### PR TITLE
drivers: flash: Add the flash driver of the stm32f1x family

### PR DESCRIPTION
**Pull request by findlayfeng**
Most of the code is copied from the stm32f0x family
Tested on stm32f103ze soc

Signed-off-by: Findlay Feng <i@fengch.me>

### CODE REVIEW

File: dts/arm/st/f1/stm32f103Xe.dtsi [hunk 1]
```diff
@@ -11,15 +11,20 @@
...
 	sram0: memory@20000000 {
 		reg = <0x20000000 DT_SIZE_K(64)>;
 	};
 
 	soc {
+		flash-controller@40022000 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(512)>;
+
+				write-block-size = <2>;
```
Comment (erwango) [inline dts/arm/st/f1/stm32f103Xe.dtsi]: `write-block-size` is common to whole series, put it in `stm32f1.dtsi`

File: dts/arm/st/f1/stm32f103X8.dtsi [hunk 2]
```diff
@@ -8,15 +8,18 @@
...
 	sram0: memory@20000000 {
 		reg = <0x20000000 DT_SIZE_K(20)>;
 	};
 
 	soc {
+
+		flash-controller@40022000 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(64)>;
+			};
```
Comment (erwango) [inline dts/arm/st/f1/stm32f103X8.dtsi]: Add `erase-block-size = <1024>;`

File: dts/arm/st/f1/stm32f103Xb.dtsi [hunk 3]
```diff
@@ -11,14 +11,18 @@
...
 	sram0: memory@20000000 {
 		reg = <0x20000000 DT_SIZE_K(20)>;
 	};
+
 	soc {
+
+		flash-controller@40022000 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(128)>;
+			};
```
Comment (erwango) [inline dts/arm/st/f1/stm32f103Xb.dtsi]: Add `erase-block-size = <1024>;`

File: dts/arm/st/f1/stm32f107Xc.dtsi [hunk 4]
```diff
@@ -8,11 +8,15 @@
...
-
 	sram0: memory@20000000 {
 		reg = <0x20000000 DT_SIZE_K(64)>;
 	};
+
+	soc {
+		flash-controller@40022000 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(256)>;
+			};
```
Comment (erwango) [inline dts/arm/st/f1/stm32f107Xc.dtsi]: Add `erase-block-size = <2048>;`

File: drivers/flash/flash_stm32f1x.c [hunk 5]
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2017 BayLibre, SAS
```
Comment (erwango) [inline drivers/flash/flash_stm32f1x.c]: Add your own copyright

File: drivers/flash/flash_stm32.c [hunk 6]
```diff
@@ -305,14 +314,17 @@ static int stm32_flash_init(struct device *dev)
 #if defined(CONFIG_SOC_SERIES_STM32L4X) || \
 	defined(CONFIG_SOC_SERIES_STM32F0X) || \
 	defined(CONFIG_SOC_SERIES_STM32F3X) || \
-	defined(CONFIG_SOC_SERIES_STM32G0X)
+	defined(CONFIG_SOC_SERIES_STM32G0X) || \
+	defined(CONFIG_SOC_SERIES_STM32F1X)
```
Comment (erwango) [inline drivers/flash/flash_stm32.c]: Move F1 bewteen F0 and F3

File: drivers/flash/flash_stm32.c [hunk 7]
```diff
@@ -305,14 +314,17 @@ static int stm32_flash_init(struct device *dev)
...
 	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
 
 	/*
 	 * On STM32F0, Flash interface clock source is always HSI,
 	 * so statically enable HSI here.
 	 */
-#if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32F3X)
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32F3X) || \
```
Comment (erwango) [inline drivers/flash/flash_stm32.c]: Move F1 between F0 and F3

File: drivers/flash/flash_stm32.h [hunk 8]
```diff
@@ -13,6 +13,7 @@
 #if defined(CONFIG_SOC_SERIES_STM32L4X) || \
 	defined(CONFIG_SOC_SERIES_STM32F0X) || \
 	defined(CONFIG_SOC_SERIES_STM32F3X) || \
```
Comment (erwango) [inline drivers/flash/flash_stm32.h]: Move F1 between F0 and F3

**Review by erwango — CHANGES_REQUESTED** [review]
Please also update nucleo_f103rb board in order to have a board upstream on which we can use and test this.
I propose the following configuration:
```
&flash0 {
	/*
	 * For more information, see:
	 * http://docs.zephyrproject.org/latest/guides/dts/index.html#flash-partitions
	 */
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		/* Set 8Kb of storage at the end of the 128Kb of flash */
		storage_partition: partition@1e000 {
			label = "storage";
			reg = <0x0001e000 0x00002000>;
		};
	};
};
```

Also, update `samples/drivers/flash_shell/sample.yaml` to have the code compiled in CI.

EDIT: `samples/drivers/flash_shell` tested ok on nucleo_f103rb with above section and requested changes.

**Comment by ioannisg** [comment]
Will we be able to get this into 2.0?

**Comment by erwango** [comment]
@findlayfeng, any plan for this work?

**Review by erwango — APPROVED** [review]

File: dts/arm/st/f1/stm32f103X8.dtsi [hunk 9]
```diff
@@ -8,15 +8,19 @@
...
 	sram0: memory@20000000 {
 		reg = <0x20000000 DT_SIZE_K(20)>;
 	};
 
 	soc {
+
+		flash-controller@40022000 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(64)>;
+				erase-block-size = <DT_SIZE_K(1)>;
```
Comment (nvlsianpu) [inline dts/arm/st/f1/stm32f103X8.dtsi]: Has this device regular flash page layout?

File: dts/arm/st/f1/stm32f103Xb.dtsi [hunk 10]
```diff
@@ -11,14 +11,18 @@
...
-
 	sram0: memory@20000000 {
 		reg = <0x20000000 DT_SIZE_K(20)>;
 	};
+
 	soc {
+		flash-controller@40022000 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(128)>;
+				erase-block-size = <DT_SIZE_K(1)>;
```
Comment (nvlsianpu) [inline dts/arm/st/f1/stm32f103Xb.dtsi]: Has this device regular flash page layout?

File: dts/arm/st/f1/stm32f103X8.dtsi (hunk 9 above)
Comment (findlayfeng) [reply dts/arm/st/f1/stm32f103X8.dtsi]: regular?

Comment (nvlsianpu) [reply dts/arm/st/f1/stm32f103X8.dtsi]: == each flash page has the same size.

Comment (findlayfeng) [reply dts/arm/st/f1/stm32f103X8.dtsi]: According to the manual there are two sizes, small and medium capacity is 1k large capacity and Internet products are 2k.

Comment (nvlsianpu) [reply dts/arm/st/f1/stm32f103X8.dtsi]: looks like almost regular, but part of the flash mapped here (the main memory) has unified page size, so looks like this is OK

**Review by nvlsianpu — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**